server/neptune/temporalworker/job/stream_handler.go

```diff
@@ -29,78 +29,75 @@ func NewTestStreamHandler(
-	streamCloserFn := func() {
-		close(streamChan)
-	}
-
-	return &StreamHandler{
-		JobOutput:        streamChan,
-		Store:            jobStore,
-		ReceiverRegistry: receiverRegistry,
-		LogFilter:        logFilter,
-		Logger:           logger,
-	}, streamCloserFn
-}
-
 type StreamHandler struct {
-	JobOutput        chan *OutputLine
 	Store            Store
+	wg               sync.WaitGroup
 	ReceiverRegistry ReceiverRegistry
 	LogFilter        filter.LogFilter
 	Logger           logging.Logger
 }
 
-func (s *StreamHandler) Stream(jobID string, msg string) {
-	s.JobOutput <- &OutputLine{
-		JobID: jobID,
-		Line:  msg,
-	}
-}
-
-func (s *StreamHandler) Handle() {
-	for msg := range s.JobOutput {
-		// Filter out log lines from job output
-		if s.LogFilter.ShouldFilterLine(msg.Line) {
-			continue
+// RegisterJob returns a channel and creates a receiver go routine
+// that reads from the associated channel.
+//
+// New resources are created each time this is called so it is up to the consumer to ensure
+// this is not being called more than once per job id.
+//
+// Note: Cleanup will be blocked until these routines are complete, therefore
+// it is essential these channels are closed or we risk deadlock on shutdown
+func (s *StreamHandler) RegisterJob(id string) chan string {
+	jobOutput := make(chan string)
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		for line := range jobOutput {
+			s.handle(&OutputLine{
+				JobID: id,
+				Line:  line,
+			})
 		}
+	}()
 
-		s.ReceiverRegistry.Broadcast(*msg)
-
-		// Append new log to the output buffer for the job
-		err := s.Store.Write(context.Background(), msg.JobID, msg.Line)
-		if err != nil {
-			s.Logger.Warn(fmt.Sprintf("appending log: %s for job: %s: %v", msg.Line, msg.JobID, err))
-		}
-	}
+	return jobOutput
 }
 
-// Activity context since it's called from within an activity
+// CloseJob closes any receivers within the registry, this exists outside of RegisterJob
+// because atm our definition of job is conflated.  In certain cases jobs are mapped to individual operations (ie. TerraformPlan/TerraformApply)
+// however, in other cases they represent a series of steps in a group (plan group -> [init, scripts, plan])
+// in order to support streaming all logs for a group to the same job id, we call this after all the activities of a group have occurred.
+// This actually is kind of broken because we don't checkpoint these across activities, so on shutdown we would lose logs for previous steps.
+// TODO: we need to rethink this a bit and create clear definitions for our data models.
 func (s *StreamHandler) CloseJob(ctx context.Context, jobID string) error {
 	s.ReceiverRegistry.Close(ctx, jobID)
 	return s.Store.Close(ctx, jobID, Complete)
 }
 
 func (s *StreamHandler) CleanUp(ctx context.Context) error {
+	// Should we timeout here? We do risk a deadlock without this however adding a timeout also risks
+	// contention with the worker stop timeout and therefore dropped logs
+	s.wg.Wait()
```

Comment:
Based on my understanding, this wait group should resolve once all the channels created in `RegisterJob` is closed right? I'm curious what happens when the operation in progress is forced to close after the timeout in `terminateOnCtxCancellation()`. Does the [scanner](https://github.com/lyft/atlantis/blob/123f5811ea67106be2f6655d935a86a2bc38fac9/server/neptune/workflows/activities/terraform.go#L363-L372) function exit and allow the channel to be closed? If not, we could run into a deadlock right since `CleanUp` is called in the main thread on server shutdown right?

Reply:
if the operation is forced to close by us signaling it then we will proceed as normal and close the activity.  If the operation is killed by kube, then it's unclear.  This shouldn't happen though given our large grace period.

Reply:
I'm not following, can you explain why we would face contention with the worker stop timeout?

Reply:
discussed offline, agree to wrap statement with logs for now